src/async_wrap.cc

```diff
@@ -199,7 +203,11 @@ void AsyncWrap::EmitTraceEventBefore() {
 }
 
 
-void AsyncWrap::EmitBefore(Environment* env, double async_id) {
+void AsyncWrap::EmitBefore(Environment* env, double async_id,
+    v8::Local<v8::Object> resource) {
+  v8::Local<v8::Context> context = env->isolate()->GetCurrentContext();
+  context->SetEmbedderData(CURRENT_RESOURCE_FIELD, resource);
```

Comment:
async_hooks are currently implemented as an per-Environment, not a per-Isolate or per-Context thing.

Reply:
I’d assume we want it to be a per-Isolate thing then, not per-Context like this currently is?

Reply:
In any case, I’d prefer not to have this PR change it only halfway. If we do switch towards per-Isolate state (which I agree is a good idea), we should move all of async_hooks over to that

Reply:
No, we'll switch towards per-Context for the `PromiseHook` as well. The fact that this is Isolate-wide was a design mistake on our end.

Reply:
Can you explain why? It seems to make tracking cross-Context async state harder?

Reply:
@mcollina Using embedder fields doesn’t seem simpler than storing it on the `Environment` … but I guess that depends on Benedikt’s reasoning here

Reply:
As @bmeurer already mentioned, we eventually want to reach a state where we have the hooks per context and not per isolate. Storing on the isolate also opens up potential optimization in the future, e.g. if we expose a JS function that can access embedder slots.

Reply:
erm. yeah. :)

Reply:
That seems like an odd choice on V8’s side, since the µtask queue is also per-Isolate … unless you also want to migrate that?

But even so, I think for Node.js per-Environment would still make more sense?